docs/azure_kms.md

```diff
@@ -1,5 +1,404 @@
+```
+
+### **Create a new Key Vault, with HSM enabled**
+
+Next we will create a new Key Vault within our newly created resource group.
+
+```sh
+az keyvault create \
+    --name KEYVAULT_NAME \
+    --resource-group RESOURCE_GROUP \
+    --sku "premium"
+```
+
+Example:
+
+```sh
+az keyvault create --name "sigy" --resource-group "sigy" --sku "premium"
+```
+
+The `--sku` argument must be set to premium if you want to have your keys stored in a HSM.
+
+### **Create a Service Principal for authentication**
+
+This document describes a service principal creation flow and a backend configuration for a key-based authentication. Alternatively, you can use client secret-based authentication but it's not recommended.
+
+You will need `openssl` to manipulate keys and certificates.
+
+#### **Generate a client certificate**
+
+```sh
+openssl req -newkey rsa:4096 -nodes -keyout "service-principal.key" -out "service-principal.csr"
+```
+
+#### **Sign a certificate**
+
+```sh
+openssl x509 -signkey "service-principal.key" -in "service-principal.csr" -req -days 365 -out "service-principal.crt"
+```
+
+Now you can safely delete the request file (`.csr`).
+
+#### **Create a Service Principal for authentication**
+
+Next we need to create a "service principal" resource (also known as a "service account" or a "App Registration").
+
+This is the credential that allows Signatory to authenticate and work with the Azure Key Vault service.
+
+```sh
+az ad sp create-for-rbac \
+    -n APP_NAME
+    --cert "@CERT_FILE"
+```
+
+Example:
+
+```sh
+az ad sp create-for-rbac -n "signatory" --cert "@service-principal.crt"
+```
+
+Example output:
+
+```sh
+Changing "sigy" to a valid URI of "http://sigy", which is the required format used for service principal names
+Certificate expires 2020-10-24 16:00:14+00:00. Adjusting SP end date to match.
+Creating a role assignment under the scope of "/subscriptions/be273d20-6dc1-4bbc-ab26-15d082cca908"
+  Retrying role assignment creation: 1/36
+  Retrying role assignment creation: 2/36
+  Retrying role assignment creation: 3/36
+  Retrying role assignment creation: 4/36
+{
+  "appId": "d5ccc5ea-8a1f-4dc1-9673-183a4c85e280",
+  "displayName": "sigy",
+  "name": "http://sigy",
+  "password": null,
+  "tenant": "50c46f11-1d0a-4c56-b468-1bcb03a8f69e"
+}
+```
+
+**Note:** If running signatory in docker then the service-principal.crt should be within the docker
+
+#### **Create a PKCS #12 file (the Microsoft way)**
+
+This method uses [PKCS #12](https://en.wikipedia.org/wiki/PKCS_12) container file for keeping a self-signed certificate along with the corresponding private key in one bundle. The certificate's hash is needed for the authentication.
+
+```sh
+openssl pkcs12 -export -out "service-principal.pfx" -inkey "service-principal.key" -in "service-principal.crt"
+```
+
+Now you can safely delete the certificate file (`.csr`).
+
+#### **Get certificate hash from AD (the rational way).**
```

Comment:
Throwing a lot of shade here lol, this may want to be removed so people don't think we hate Microsoft too much